server-common/src/main/java/org/apache/kafka/queue/KafkaEventQueue.java

```diff
@@ -278,21 +280,24 @@ private void handleEvents() {
                         remove(toRun);
                         continue;
                     }
-                    if (awaitNs == Long.MAX_VALUE) {
-                        try {
+
+                    long startIdleNs = time.nanoseconds();
+                    try {
+                        if (awaitNs == Long.MAX_VALUE) {
                             cond.await();
-                        } catch (InterruptedException e) {
-                            log.warn("Interrupted while waiting for a new event. " +
-                                "Shutting down event queue");
-                            interrupted = true;
-                        }
-                    } else {
-                        try {
+                        } else {
                             cond.awaitNanos(awaitNs);
-                        } catch (InterruptedException e) {
-                            log.warn("Interrupted while waiting for a deferred event. " +
-                                "Shutting down event queue");
-                            interrupted = true;
+                        }
+                    } catch (InterruptedException e) {
+
+                        log.warn("Interrupted while waiting for a {} event. " +
+                                "Shutting down event queue", (awaitNs == Long.MAX_VALUE) ? "new" : "deferred");
+                        interrupted = true;
+                    } finally {
+                        if (idleTimeCallback != null) {
+                            long idleNs = Math.max(time.nanoseconds() - startIdleNs, 0);
+                            long idleMs = TimeUnit.NANOSECONDS.toMillis(idleNs);
```

Comment:
Please use `time.milliseconds` since you are converting to milliseconds anyways.